packages/mongodb-artifact-generator/src/release-notes/summarizeReleaseArtifacts.ts

```diff
@@ -0,0 +1,154 @@
+import { stripIndents, html } from "common-tags";
+import {
+  GenerateChatCompletion,
+  makeGenerateChatCompletion,
+  systemMessage,
+  userMessage,
+} from "../chat";
+import { ReleaseArtifact, releaseArtifactIdentifier } from "./projects";
+import { RunLogger } from "../runlogger";
+import { PromisePool } from "@supercharge/promise-pool";
+
+export function safeFileName(fileName: string) {
+  return fileName.replace(/[/\\?%*:|"<>]/g, "-");
+}
+
+export type SummarizeReleaseArtifactArgs = {
+  logger?: RunLogger;
+  generate?: GenerateChatCompletion;
+  projectDescription: string;
+  artifact: ReleaseArtifact;
+};
+
+export async function summarizeReleaseArtifact({
+  logger,
+  generate,
+  projectDescription,
+  artifact,
+}: SummarizeReleaseArtifactArgs) {
+  generate = generate ?? makeGenerateChatCompletion();
+  const chatTemplate = [
+    systemMessage(stripIndents`
+      Your task is to analyze a provided artifact associated with a software release and write a succinct summarized description. Artifacts may include information from task tracking software like Jira, source control information like Git diffs and commit messages, or other sources.
+
+      Your summary should be a brief, high-level description of the artifact's contents and purpose. It should avoid technical details and focus on the artifact's significance and relevance to the project. The goal is to provide a clear, concise overview that can be used to generate a change log entry for the release.
+
+      The user may prepend the artifact with additional style guide information or other metadata. This section is denoted by frontmatter enclosed in triple dashes (---). Do not mention this frontmatter in the summary but do follow its guidance.
+
+      Limit the summary length to a maximum of 200 words.
+
+      <Section description="A description of the software project this release is for">
+        ${projectDescription}
+      </Section>
+    `),
+    userMessage(createUserPromptForReleaseArtifact(artifact)),
+  ];
+  logger?.appendArtifact(
+    `chatTemplates/${safeFileName(releaseArtifactIdentifier(artifact))}`,
+    chatTemplate
+      .map((m) => {
+        switch (m.role) {
+          case "system":
+            return `<SystemMessage>\n${m.content}\n</SystemMessage>`;
+          case "user":
+            return `<UserMessage>\n${m.content}\n</UserMessage>`;
+        }
+      })
+      .join("\n")
+  );
+  logger?.logInfo(
+    `[summarizeReleaseArtifact chatTemplate] ${JSON.stringify(chatTemplate)}`
+  );
+
+  const output = await generate(chatTemplate);
+  if (!output) {
+    throw new Error("Something went wrong while generating the summary ☹️");
+  }
+  return output;
+}
+
+function frontmatter(
```

Comment:
can you use https://www.npmjs.com/package/front-matter instead of writing this out?

Reply:
That lib parses YAML frontmatter into objects. This goes the other direction, i.e. turning data into a frontmatter string.

Reply:
then maybe use yaml.stringiy()? we do something similar in the ingestion code